src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs

```diff
@@ -761,7 +784,7 @@ private static unsafe string JoinCore<T>(char* separator, int separatorLength, I
             }
         }
 
-        private static unsafe string JoinCore(char* separator, int separatorLength, string?[] value, int startIndex, int count)
+        private static unsafe string JoinCore(char* separator, int separatorLength, ReadOnlySpan<string?> value, int startIndex, int count)
```

Comment:
Now that this takes a span, are startIndex and count needed?  Callers can slice if necessary.  And this implementation will then be faster as it'll be able to iterate over the full size of the span and bounds checks will be removed.

Reply:
In `public static unsafe string Join(char separator, string?[] value, int startIndex, int count)` would i do `value.AsSpan(startIndex, count)`?

Reply:
Yes.

And now that I look at the implementation more carefully to answer your question, I'm realizing the change still needs further work, anyway.  Spans can never be null, so the change to the type of `value` is removing a null check that should have been there for the `public static unsafe string Join(string? separator, string?[] value, int startIndex, int count)` overload.  By doing the AsSpan, we'd similarly need to move the argument validation for startIndex and count to the relevant entry point(s) that need the arguments validated.

Reply:
I'll move this and other public overloads to pass down the value.AsSpan, cleanup the JoinCore implementation, and benchmark again to verify there are no regressions. Do we want to move the argument validation of startIndex and count to the public overloads even though `value.ToSpan(startIndex, count)` is doing the same validations (with less helpful error messages)?

Regarding the missing test there is already `AssertExtensions.Throws<ArgumentNullException>("value", () => string.Join('|', (string[])null, 0, 0));` I'll RC why this wasn't failing.